custom/icds_reports/filters.py

```diff
@@ -35,40 +35,66 @@ def loc_to_json(loc, project):
+
+    if user_has_all_location_access:
+        accessible_root_locations = SQLLocation.root_locations(domain)
+    else:
+        accessible_root_locations = (user_location
+                                     .get_ancestors(include_self=True)
+                                     .filter(parent_id__isnull=True))
 
-    locations = SQLLocation.root_locations(domain)
     if not show_test:
-        locations = [
-            loc for loc in locations if loc.metadata.get('is_test_location', 'real') != 'test'
-        ]
+        accessible_root_locations = [loc for loc in accessible_root_locations
+                                     if loc.metadata.get('is_test_location', 'real') != 'test']
 
-    loc_json = [loc_to_json(loc, project) for loc in locations]
+    loc_json = [loc_to_json(loc) for loc in accessible_root_locations]
 
     # if a location is selected, we need to pre-populate its location hierarchy
     # so that the data is available client-side to pre-populate the drop-downs
     if selected_loc_id:
+        if not user_has_all_location_access:
+            accessible_location_ids = _get_accessible_location_ids(user_location)
+        else:
+            accessible_location_ids = []
+
         selected = SQLLocation.objects.get(
             domain=domain,
             location_id=selected_loc_id
         )
 
-        lineage = selected.get_ancestors()
-
-        parent = {'children': loc_json}
-        for loc in lineage:
-            children = loc.child_locations()
-            # find existing entry in the json tree that corresponds to this loc
-            try:
-                this_loc = [k for k in parent['children'] if k['uuid'] == loc.location_id][0]
-            except IndexError:
-                # if we couldn't find this location the view just break out of the loop.
-                # there are some instances in viewing archived locations where we don't actually
-                # support drilling all the way down.
+        json_at_level = loc_json  # json in which we should find the ancestor in iteration
+        for ancestor in selected.get_ancestors():  # this would start with top level ancestor first
+            ancestor_loc_dict_in_json = _get_ancestor_loc_dict(json_at_level, ancestor)
+
+            # could not find the ancestor at the level,
+            # user should not have reached at this point to try and access an ancestor that is not permitted
+            if ancestor_loc_dict_in_json is None:
                 break
-            this_loc['children'] = [loc_to_json(loc, project) for loc in children]
-            parent = this_loc
 
+            child_locations = ancestor.get_children()
+
+            ancestor_loc_dict_in_json['children'] = [
+                loc_to_json(loc) for loc in child_locations
+                if loc.metadata.get('is_test_location', 'real') != 'test'
+                and (user_has_all_location_access or loc.location_id in accessible_location_ids)
```

Comment:
I am not sure if that would be that straight forward, but probably would be clear on giving it a shot.

but just weighing in, If we change that method to get accessible locations instead of just ids, we would first need to set some structure to that, group them by levels or something and then again do some manipulation to filter out what is relevant for the current search to return. If we succeed at that, can't say if it will still be better.
Additionally, It's possible we won't even need all the locations fetched, for ex: if a user has access to a state, we would be fetching all locations under that state, and currently if the user is selecting say a district, we don't really need location data below that district. 
Additionally, Location search has been reported to be a pain on ICDS and there is a ticket to work on that, 
it slows down extremely, at

https://github.com/dimagi/commcare-hq/blob/8f913988b5c0d1c2e8092c2ea2d1d6915a18c0d5/corehq/apps/locations/models.py#L268-L272

so fetching data for all assigned locations is heavy on ICDS, so would be worth avoiding it. i guess what we are doing here is faster.

this method is quite a duplicate of `load_locs_json` in `locations/utils` which follows the same pattern, so if this an improvement, would be nice to extend it to that as well.

Reply:
agree with Manish